ts/input/tex/textmacros/TextParser.ts

```diff
@@ -0,0 +1,198 @@
+   */
+  public text: string;
+
+  /**
+   * Saved stack environments for processing braces
+   */
+  public envStack: EnvList[];
+
+  /**
+   * The scriptlevel of this text-mode material
+   */
+  public level: number | string | undefined;
+
+  /**
+   * The accumulated MmlNodes generated by parsing the text-mode material
+   */
+  protected nodes: MmlNode[];
+
+  /**
+   * Short-hand for obtaining the saved TexParser
+   */
+  public get texParser() {
+    return this.configuration.packageData.textmacros.texParser;
+  }
+
+  /**
+   * @override
+   * @constructor
+   */
+  constructor(text: string, env: EnvList, configuration: ParseOptions, level?: number | string) {
+    super(text, env, configuration);
+    this.level = level;
+  }
+
+  /**
+   * Make sure we only return one element (wrap multiple ones in an mrow or mstyle, as needed).
+   *
+   * @override
+   */
+  public mml() {
+    return (this.level != null ?
+            this.create('node', 'mstyle', this.nodes, {displaystyle: false, scriptlevel: this.level}) :
+            this.nodes.length === 1 ? this.nodes[0] : this.create('node', 'inferredMrow', this.nodes));
+  }
+
+  /**
+   * @override
+   */
+  public Parse() {
+    this.text = '';
+    this.nodes = [];
+    this.envStack = [];
+    super.Parse();
+  }
+
+  /**
+   * Creates an mtext element for the saved text and pushes that onto the node list
+   */
+  public saveText() {
+    if (this.text) {
+      const text = this.create('token', 'mtext', {}, this.text);
+      if (this.stack.env.mathvariant) {
+        NodeUtil.setAttribute(text, 'mathvariant', this.stack.env.mathvariant);
+      }
+      this.text = '';
+      this.Push(text);
+    }
+  }
+
+  /**
+   * @override
+   */
+  public Push(mml: MmlNode | StackItem) {
+    if (this.text) {
+      this.saveText();
+    }
+    if (mml instanceof StopItem) {
+      return super.Push(mml);
+    }
+    if (mml instanceof StyleItem) {
+      this.stack.env.mathcolor = this.stack.env.color;
+      return;
+    }
+    if (mml instanceof AbstractMmlNode) {
+      this.addAttributes(mml);
+      this.nodes.push(mml);
+    }
+  }
+
+  /**
+   * Push some math into the node list, adding mathsize and mathcolor
+   *   if specified in the environment.
+   *
+   * @param {MmlNode} mml   The math nodes to push
+   */
+  public PushMath(mml: MmlNode) {
+    const env = this.stack.env;
+    for (const name of ['mathsize', 'mathcolor']) {
+      if (env[name] && !mml.attributes.getExplicit(name)) {
+        if (!mml.isToken && !mml.isKind('mstyle')) {
+          mml = this.create('node', 'mstyle', [mml]);
+        }
+        NodeUtil.setAttribute(mml, name, env[name]);
+      }
+    }
+    this.nodes.push(mml);
+  }
+
+  /**
+   * Add mathsize, mathcolor, and mathvariant to token nodes,
+   *   if they are specified in the environment
+   *
+   * @param {MmlNode} mml   The node to adjust
+   */
+  public addAttributes(mml: MmlNode) {
+    const env = this.stack.env;
+    if (mml.isToken) {
```

Comment:
Swap order and start with `if (!mml.isToken) return;`

Reply:
Done